prov/rxm/src/rxm_ep.c

```diff
@@ -89,22 +89,100 @@ static int rxm_match_unexp_msg_tagged(struct dlist_entry *item, const void *arg)
 		rxm_match_tag(attr->tag, attr->ignore, unexp_msg->tag);
 }
 
-static void rxm_mr_buf_close(void *pool_ctx, void *context)
+static inline int rxm_mr_buf_reg(struct rxm_ep *rxm_ep, void *addr, size_t len, void **context)
 {
-	/* We would get a (fid_mr *) in context but it is safe to cast it into (fid *) */
-	fi_close((struct fid *)context);
+	int ret = FI_SUCCESS;
+	struct fid_mr *mr;
+	struct rxm_domain *rxm_domain = container_of(rxm_ep->util_ep.domain,
+						     struct rxm_domain, util_domain);
+
+	*context = NULL;
+	if (rxm_ep->msg_mr_local) {
+		struct fid_domain *msg_domain =
+			(struct fid_domain *)rxm_domain->msg_domain;
+
+		ret = fi_mr_reg(msg_domain, addr, len,
+				FI_SEND | FI_RECV | FI_READ | FI_WRITE,
+				0, 0, 0, &mr, NULL);
+		*context = mr;
+	}
+
+	return ret;
+}
+
+#define rxm_fill_rx_buf_default(buf, rxm_ep, mr_desc)	\
+	do {						\
+		(buf)->ep = rxm_ep;			\
+		(buf)->hdr.desc = mr_desc;		\
+	} while (0)
+
+#define rxm_fill_tx_msg_buf_default(buf, rxm_ep, mr_desc)	\
+	do {							\
+		(buf)->pkt.ctrl_hdr.type = ofi_ctrl_data;	\
+		(buf)->pkt.ctrl_hdr.version = OFI_CTRL_VERSION;	\
+		(buf)->pkt.hdr.op = ofi_op_msg;			\
+		(buf)->pkt.hdr.version = OFI_OP_VERSION;	\
+		(buf)->hdr.desc = mr_desc;			\
+	} while (0)
+
+#define rxm_fill_tx_tagged_buf_default(buf, rxm_ep, mr_desc)	\
+	do {							\
+		(buf)->pkt.ctrl_hdr.type = ofi_ctrl_data;	\
+		(buf)->pkt.ctrl_hdr.version = OFI_CTRL_VERSION;	\
+		(buf)->pkt.hdr.op = ofi_op_tagged;		\
+		(buf)->pkt.hdr.version = OFI_OP_VERSION;	\
+		(buf)->hdr.desc = mr_desc;			\
+	} while (0)
+
+#define rxm_fill_tx_ack_buf_default(buf, rxm_ep, mr_desc)	\
+	do {							\
+		(buf)->pkt.ctrl_hdr.type = ofi_ctrl_ack;	\
+		(buf)->pkt.ctrl_hdr.version = OFI_CTRL_VERSION;	\
+		(buf)->pkt.hdr.op = ofi_op_msg;			\
+		(buf)->pkt.hdr.version = OFI_OP_VERSION;	\
+		(buf)->hdr.desc = mr_desc;			\
+	} while (0)
+
+#define RXM_DEFINE_BUF_REG(op_prefix, type_suffix)						\
+static int rxm_ ## op_prefix ## type_suffix ## _buf_reg(void *pool_ctx, void *addr,		\
+							 size_t len, void **context)		\
+{												\
+	struct rxm_ep *rxm_ep = (struct rxm_ep *)pool_ctx;					\
+	size_t i, entry_sz = fi_get_aligned_sz(rxm_ep->rxm_info->tx_attr->inject_size +		\
+					       sizeof(struct rxm_ ## op_prefix ## _buf) +	\
+					       sizeof(struct util_buf_footer), 16);		\
+	int ret;										\
+	struct rxm_ ## op_prefix ## _buf *buf;							\
+	void *mr_desc;										\
+												\
+	ret = rxm_mr_buf_reg(rxm_ep, addr, len, context);					\
+	if (ret)										\
+		return ret;									\
+												\
+	mr_desc = (*context != NULL) ? fi_mr_desc((struct fid_mr *)*context) : NULL;		\
+												\
+	for (i = 0; i < rxm_ep->msg_info->op_prefix ## _attr->size; i++) {			\
+		buf = (struct rxm_ ## op_prefix ## _buf *)((char *)addr + i * entry_sz);	\
+		rxm_fill_ ## op_prefix ## type_suffix ## _buf_default(buf, rxm_ep, mr_desc);	\
+	}											\
+												\
+	return FI_SUCCESS;									\
```

Comment:
I'd prefer a function here as well, rather than something auto-generated.

Reply:
Ok, I agree, it would be nice